### PR TITLE
Fix empty verification bottom sheet

### PR DESCRIPTION
**Pull request by bmarty**
## Type of change

- [ ] Feature
- [x] Bugfix
- [ ] Technical
- [ ] Other :

## Content

When the User with `otherUserId` was not known by the SDK, it was not possible to verify them.

First commit https://github.com/vector-im/element-android/pull/7130/commits/6d2a9ec9d5535b04ec5e9bcafc20da0e27e84c79 is the first part of the fix. `otherUserId` was null because retrieved from a null `MatrixItem`.
Second commit https://github.com/vector-im/element-android/pull/7130/commits/92d73912323f7b6ce8f4fff0c22941cf44ce2590 ensure that the `MatrixItem` cannot be null.

The 2nd commit should be sufficient to fix the issue, but the first commit make thing clearer IMHO.

## Motivation and context

Be able to Verify every users.

## Screenshots / GIFs

|Before|After|
|-|-|
|imageimage|imageimage|

## Tests

- Step 1
- Step 2
- Step ...

## Tested devices

- [ ] Physical
- [ ] Emulator
- OS version(s):

## Checklist

- [ ] Changes has been tested on an Android device or Android emulator with API 21
- [ ] UI change has been tested on both light and dark themes
- [ ] Accessibility has been taken into account. See https://github.com/vector-im/element-android/blob/develop/CONTRIBUTING.md#accessibility
- [ ] Pull request is based on the develop branch
- [ ] Pull request includes a new file under ./changelog.d. See https://github.com/vector-im/element-android/blob/develop/CONTRIBUTING.md#changelog
- [ ] Pull request includes screenshots or videos if containing UI changes
- [ ] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
- [ ] You've made a self review of your PR
- [ ] If you have modified the screen flow, or added new screens to the application, you have updated the test [UiAllScreensSanityTest.allScreensTest()](https://github.com/vector-im/element-android/blob/main/vector/src/androidTest/java/im/vector/app/ui/UiAllScreensSanityTest.kt#L73)

### CODE REVIEW

File: vector/src/main/java/im/vector/app/features/crypto/verification/VerificationBottomSheetViewModel.kt [hunk 1]
```diff
@@ -126,7 +128,7 @@ class VerificationBottomSheetViewModel @AssistedInject constructor(
             }
         }
 
-        val userItem = session.getUser(initialState.otherUserId)
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/crypto/verification/VerificationBottomSheetViewModel.kt:129]: A null `userItem` was the source of the issue.

File: vector/src/main/java/im/vector/app/features/crypto/verification/VerificationBottomSheetViewModel.kt [hunk 2]
```diff
@@ -160,7 +162,6 @@ class VerificationBottomSheetViewModel @AssistedInject constructor(
 
         setState {
             copy(
-                    otherUserMxItem = userItem?.toMatrixItem(),
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/crypto/verification/VerificationBottomSheetViewModel.kt:163]: is set in the constructor and updated with full data in `fetchOtherUserProfile`

File: vector/src/main/java/im/vector/app/features/crypto/verification/VerificationBottomSheetViewModel.kt [hunk 3]
```diff
@@ -183,6 +184,28 @@ class VerificationBottomSheetViewModel @AssistedInject constructor(
         }
     }
 
+    private fun fetchOtherUserProfile(otherUserId: String) {
+        session.getUser(otherUserId)?.toMatrixItem()?.let {
+            setState {
+                copy(
+                        otherUserMxItem = it
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/crypto/verification/VerificationBottomSheetViewModel.kt:191]: First change to set known data

File: vector/src/main/java/im/vector/app/features/crypto/verification/VerificationBottomSheetViewModel.kt [hunk 4]
```diff
@@ -183,6 +184,28 @@ class VerificationBottomSheetViewModel @AssistedInject constructor(
...
+            }
+        }
+        // Always fetch the latest User data
+        viewModelScope.launch {
+            tryOrNull { session.userService().resolveUser(otherUserId) }
+                    ?.toMatrixItem()
+                    ?.let {
+                        setState {
+                            copy(
+                                    otherUserMxItem = it
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/crypto/verification/VerificationBottomSheetViewModel.kt:202]: second change to set up to date data, if possible (network available + API is returning data)

File: vector/src/main/java/im/vector/app/features/crypto/verification/VerificationBottomSheetViewModel.kt [hunk 5]
```diff
@@ -216,12 +239,12 @@ class VerificationBottomSheetViewModel @AssistedInject constructor(
 
     private fun cancelAllPendingVerifications(state: VerificationBottomSheetViewState) {
         session.cryptoService()
-                .verificationService().getExistingVerificationRequest(state.otherUserMxItem?.id ?: "", state.transactionId)?.let {
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/crypto/verification/VerificationBottomSheetViewModel.kt:219]: defaulting to `""` everywhere is pretty bad...

**Comment by bmarty** [comment]
Adding @BillCarsonFr as a reviewer, since this is code related to Crypto.

**Comment by ganfra** [comment]
Probably because of the move to worker for fetching users?

**Review by BillCarsonFr — APPROVED** [review]
LGTM.
I can see that getUser() is also called in IncomingVerificationRequestHandler.
It should call resolveUser or at least default to a MatriItem with the correct userId?

**Comment by bmarty** [comment]
> Probably because of the move to worker for fetching users?

I do not think so, because even all attempts was unsuccessful. With this theory, after a while it should work. I think this is not a regression, but an old bug.

**Comment by bmarty** [comment]
> LGTM. I can see that getUser() is also called in IncomingVerificationRequestHandler. It should call resolveUser or at least default to a MatriItem with the correct userId?

Thanks, let me have a look on it.

**Comment by bmarty** [comment]
> > LGTM. I can see that getUser() is also called in IncomingVerificationRequestHandler. It should call resolveUser or at least default to a MatriItem with the correct userId?
> 
> Thanks, let me have a look on it.

**Review by onurays — APPROVED** [review]
Nice improvement. Statically reviewed, LGTM! It was annoying to fallback to `""`.

**Pull request merged**